plugin/health_check.go

```diff
@@ -0,0 +1,165 @@
+func (env *Environment) InitPluginHealthCheckJob(enable bool) {
+	// Config is set to enable. No job exists, start a new job.
+	if enable && env.pluginHealthCheckJob == nil {
+		job := newPluginHealthCheckJob(env)
+		env.pluginHealthCheckJob = job
+		job.Start()
+	}
+
+	// Config is set to disable. Job exists, kill existing job.
+	if !enable && env.pluginHealthCheckJob != nil {
+		env.pluginHealthCheckJob.Cancel()
+		env.pluginHealthCheckJob = nil
+	}
+}
+
+// Start continuously runs health checks on all active plugins, on a timer.
+func (job *PluginHealthCheckJob) Start() {
+	interval := time.Duration(HEALTH_CHECK_INTERVAL) * time.Second
+	mlog.Debug(fmt.Sprintf("Plugin health check job starting. Sending health check pings every %v seconds.", interval))
+
+	go func() {
+		defer close(job.cancelled)
+
+		ticker := time.NewTicker(interval)
+		defer func() {
+			ticker.Stop()
+		}()
+
+		for {
+			select {
+			case <-ticker.C:
+				activePlugins := job.env.Active()
+				for _, plugin := range activePlugins {
+					job.checkPlugin(plugin.Manifest.Id)
+				}
+			case <-job.cancel:
+				return
+			}
+		}
+	}()
+}
+
+// checkPlugin determines the plugin's health status, then handles the error or success case.
+func (job *PluginHealthCheckJob) checkPlugin(id string) {
+	var ap activePlugin
+	if p, ok := job.env.activePlugins.Load(id); ok {
+		ap = p.(activePlugin)
+	} else {
+		return
+	}
+
+	if _, ok := job.env.pluginHealthStatuses.Load(id); !ok {
+		job.env.pluginHealthStatuses.Store(id, newPluginHealthStatus())
+	}
+
+	sup := ap.supervisor
+	if sup == nil {
+		return
+	}
+
+	pluginErr := sup.PerformHealthCheck()
+
+	if pluginErr != nil {
+		mlog.Debug(fmt.Sprintf("Health check failed for plugin %s, error: %s", id, pluginErr.Error()))
+		job.handleHealthCheckFail(id, pluginErr)
+	} else {
+		job.handleHealthCheckSuccess(id)
+	}
+}
+
+// handleHealthCheckFail restarts or deactivates the plugin based on how many times it has failed in a configured amount of time.
+func (job *PluginHealthCheckJob) handleHealthCheckFail(id string, err error) {
+	var h *PluginHealthStatus
+	if health, ok := job.env.pluginHealthStatuses.Load(id); ok {
+		h = health.(*PluginHealthStatus)
+	} else {
+		return
+	}
+
+	// Append current failure before checking for deactivate vs restart action
+	h.failTimeStamps = append(h.failTimeStamps, time.Now())
+	h.lastError = err
+
+	if shouldDeactivatePlugin(h) {
+		h.failTimeStamps = []time.Time{}
+		h.Crashed = true
+		mlog.Debug(fmt.Sprintf("Deactivating plugin due to multiple crashes `%s`", id))
+		job.env.Deactivate(id)
+	} else {
+		mlog.Debug(fmt.Sprintf("Restarting plugin due to failed health check `%s`", id))
+		job.env.RestartPlugin(id)
+	}
+}
+
+// handleHealthCheckSuccess marks the plugin as healthy.
+func (job *PluginHealthCheckJob) handleHealthCheckSuccess(id string) {
+	job.env.UpdatePluginHealthStatus(id, func(h *PluginHealthStatus) {
+		h.Crashed = false
```

Comment:
Can this be the case? Is `checkPlugin` for a crashed plugin?

Reply:
`checkPlugin()` is the function that runs per plugin, each interval of the job. It runs for each `active` plugin at the time. If a plugin is deemed healthy, it is marked as such here.

The name `checkPlugin` could be more revealing as for what the function does, since it also has side effects like restarting the plugin.

Reply:
Sorry, my original comment was misleading and wrong.

My questions is: Does `handleHealthCheckSuccess` ever get called, when `h.Crashed = true'? I thought, if `h.Crashed = true`, the plugin will get deactivated.

Reply:
I understand your point. Originally, the plugin was not deemed healthy until it passed its first health check. At the current state of the code, the `Crashed = false` value is being [initialized](https://github.com/mattermost/mattermost-server/pull/10781/files#diff-bc32e81beea4185ff5e113fef62924c1R240) in `env.Activate()` before the first health check is run, so it is deemed healthy on activation. 

This `handleHealthSuccess()` function can then be removed, since the `Crashed` value is already set at the point of the first health check, and this function is doing nothing except setting that value. How does this sound @hanzei?

Reply:
I pushed changes, including removing this function